common/src/terraingen.rs

```diff
@@ -0,0 +1,996 @@
+            VoronoiInfo::new(Material::Andesite, -4.50, -10.50),
+            VoronoiInfo::new(Material::Olivine, -4.50, -7.50),
+            VoronoiInfo::new(Material::Andesite, -4.50, -4.50),
+            VoronoiInfo::new(Material::Gabbro, -4.50, -1.50),
+            VoronoiInfo::new(Material::Gabbro, -4.50, 1.50),
+            VoronoiInfo::new(Material::Gabbro, -4.50, 4.50),
+            VoronoiInfo::new(Material::Gabbro, -4.50, 7.50),
+            VoronoiInfo::new(Material::Gabbro, -4.50, 10.50),
+            VoronoiInfo::new(Material::Olivine, -6.00, -9.00),
+            VoronoiInfo::new(Material::Andesite, -6.00, -6.00),
+            VoronoiInfo::new(Material::Olivine, -6.00, -3.00),
+            VoronoiInfo::new(Material::Gabbro, -6.00, 0.00),
+            VoronoiInfo::new(Material::Gabbro, -6.00, 3.00),
+            VoronoiInfo::new(Material::Basalt, -6.00, 6.00),
+            VoronoiInfo::new(Material::Gabbro, -6.00, 9.00),
+            VoronoiInfo::new(Material::Andesite, -7.50, -10.50),
+            VoronoiInfo::new(Material::Olivine, -7.50, -7.50),
+            VoronoiInfo::new(Material::Olivine, -7.50, -4.50),
+            VoronoiInfo::new(Material::Gabbro, -7.50, -1.50),
+            VoronoiInfo::new(Material::Olivine, -7.50, 1.50),
+            VoronoiInfo::new(Material::Gabbro, -7.50, 4.50),
+            VoronoiInfo::new(Material::Basalt, -7.50, 7.50),
+            VoronoiInfo::new(Material::Gabbro, -7.50, 10.50),
+            VoronoiInfo::new(Material::Olivine, -9.00, -9.00),
+            VoronoiInfo::new(Material::Olivine, -9.00, -6.00),
+            VoronoiInfo::new(Material::Olivine, -9.00, -3.00),
+            VoronoiInfo::new(Material::Olivine, -9.00, 0.00),
+            VoronoiInfo::new(Material::Basalt, -9.00, 3.00),
+            VoronoiInfo::new(Material::Basalt, -9.00, 6.00),
+            VoronoiInfo::new(Material::Basalt, -9.00, 9.00),
+            VoronoiInfo::new(Material::Olivine, -10.50, -10.50),
+            VoronoiInfo::new(Material::Olivine, -10.50, -7.50),
+            VoronoiInfo::new(Material::Olivine, -10.50, -4.50),
+            VoronoiInfo::new(Material::Olivine, -10.50, -1.50),
+            VoronoiInfo::new(Material::Basalt, -10.50, 1.50),
+            VoronoiInfo::new(Material::Basalt, -10.50, 4.50),
+            VoronoiInfo::new(Material::Basalt, -10.50, 7.50),
+            VoronoiInfo::new(Material::Lava, -10.50, 10.50),
+        ];
+
+        let mut voxel_mat: Material;
+
+        let voronoi_choices = if elev < -30.0 {
+            &voronoi_choices_deep
+        } else if elev < -12.0 {
+            &voronoi_choices_low
+        } else if elev < 4.0 {
+            &voronoi_choices_med
+        } else {
+            &voronoi_choices_high
+        };
+
+        let y = na::Vector2::new(rain, temp);
+        let mut dist = na::Matrix::norm(&(voronoi_choices[0].location - y));
+        voxel_mat = voronoi_choices[0].material;
+        for vc in voronoi_choices.iter().skip(1) {
```

Comment:
I'm concerned that we're iterating through a fairly large number of values here brute-force. We do have some spare CPU to throw around in worldgen for the time being, but this feels like squandering it. Could we precompute results into an NxNxN 3D array or similar instead? That would be *much* faster.

Relatedly, it might be useful to create a benchmark for the generation of a chunk (say, the origin chunk).

Reply:
I have not done any precomputation. That would probably be something I'd need help with. I am also unsure about the amount of space required for the array. Suppose for temp, rain, and elev in the range of (-10.0, 10.0) you wanted a resolution of 0.02, yielding 1000 samples, then you'd have a billion coordinates, each storing an integer for the Material enum value. Perhaps 500 or even 200 samples would still look good, which would greatly reduce the memory cost. An alternative is to only sample for rain and temp at a small set of elevations, so the storage would be merely 1000x1000xn. 

For now, I have changed the distance calculation to give distance squared.